src/csharp/Microsoft.Spark/Interop/Ipc/ArrayBufferWriter.cs

```diff
@@ -0,0 +1,185 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Buffers
+{
+    /// <summary>
+    /// Represents a heap-based, array-backed output sink into which <typeparam name="T"/> data can be written.
+    /// </summary>
+    public class ArrayBufferWriter<T> : IBufferWriter<T>
```

Comment:
nit: consider adding a comment for where this was copied from (in case some sync-ing is required in the future).